docs/source/prometheus.rst

```diff
@@ -117,6 +125,14 @@ dask_worker_tasks
     Number of tasks at worker
 dask_worker_threads
     Number of worker threads
+dask_worker_gil_contention:
+    0-1 value representing GIL contention on worker
+
+    .. note::
+       Requires ``gilknocker`` to be installed, and 
+       ``distributed.admin.system-monitor.gil-contention.enabled``
+       configuration to be set.
```

Comment:
I'm curious, is this on by default?

Reply:
No it is [off by default](https://github.com/dask/distributed/blob/6cab0e25a0044e68641c88664df315526b60a7c6/distributed/distributed.yaml#L313)

Reply:
I'll express a desire to have this eventually (in the next couple of months) on by default, at least in Coiled.  There are a few ways to get there.  For example we could ...

1. Turn the configuration on by default in Dask config (still gated on `gil_knocker` being installed)
2. Have Coiled depend on `gil_knocker` and plan on package-sync to distribute that package

Some questions if we wanted to go further than that:

1.  Are there negative side effects to running gil_knocker in practice?  Is it reasonable to not want it to be on?  The cost here is less than a percent?
2. How difficult is it to distribute gil_knocker?  Do we have conda packages on conda-forge?  Are wheels easy to build?

Reply:
> Are there negative side effects to running gil_knocker in practice? Is it reasonable to not want it to be on? The cost here is less than a percent?

We've noted in some [benchmarks are affected](https://github.com/coiled/coiled-runtime/actions/runs/4402658596) using the default `1 millisecond` polling. In running some rough bench-marking this 1ms interval appears to reflect what `py-spy` gives in most test cases, and bumping it to 5ms reduces any detectable performance impact AFAICT but is much less representative of actual GIL contention. 

In those benchmarks, some are not affected (but most have at least some impact) with the worst offender being `test_h2o`:
![image](https://user-images.githubusercontent.com/13764397/225003926-350e4542-36d0-4cc2-9f4f-b65d1cd6af5a.png)


For example, with 5ms, 30% contention is likely much higher, in the 60-90% maybe. But this varies wildly depending on the type of program being ran and how frequent/little/long it holds the GIL in the first place; therefore becoming more of a rough indicator to which the user could then opt to increase polling interval to get a closer idea of actual contention. Maybe this is a 'safe' option, default on with a 5ms polling interval?

---


> How difficult is it to distribute gil_knocker? Do we have conda packages on conda-forge? Are wheels easy to build?

[Building wheels is pretty easy](https://github.com/milesgranger/gilknocker/blob/main/.github/workflows/CI.yml) with [a decent amount of support already](https://pypi.org/project/gilknocker/#files) and can be expanded slightly to use a different project's build I maintain [cramjam](https://github.com/milesgranger/pyrus-cramjam) which builds wheels for basically every possible platform. 

However, if having conda packages is a requirement, it appears such [builds need to be replicated on its feedstock](https://github.com/conda-forge/cramjam-feedstock/blob/main/recipe/meta.yaml) which is slightly annoying, especially when a package like cramjam or gilknocker has no dependencies.

Reply:
Had some other thoughts for lessening the effect a high monitoring interval might have over here: https://github.com/milesgranger/gilknocker/issues/9

Reply:
After updating the changes in `gilknocker` and [re-running the same A/B benchmarks](https://github.com/coiled/coiled-runtime/actions/runs/4426271537), it appears the effect is much less and could be lowered still by either allowing for larger gaps in sampling frequency and/or bumping the polling interval to a lower frequency. Here are some A/B benchmarks after the updates using 1ms frequency (same frequency as above before updates)

![image](https://user-images.githubusercontent.com/13764397/225558161-9a7f177d-abdb-4017-a6b1-7572c26b4e2e.png)

Reply:
I would also like to see this on by default if gilknocker is installed. I wonder if the interval could be configurable and then set the default to some value that doesn't impact performance and document how to increase it for better accuracy with the appropriate performance warnings?

Reply:
Sorry for the noise here.. but with gilknocker==0.4.0 [in the updated benchmarks](https://github.com/coiled/coiled-runtime/actions/runs/4437855278) there appears to be no noticeable impact on performance with a 1ms polling interval (the default in the config) The above (outdated) graph was with gilknocker==0.3.0 w/ 1ms polling.